src/plugins/intel_gpu/src/graph/network.cpp

```diff
@@ -1508,4 +1516,57 @@ void network::assign_variables_memories(variables_states_map &&variables_memorie
     }
 }
 
+void network::assign_variables_memories() {
+    for (auto primitive : _variable_state_primitives) {
+        if (const auto& memory_state_primitive = std::dynamic_pointer_cast<memory_state::variable>(primitive)) {
+            auto it = _variables_states.find(memory_state_primitive->variable_id());
+            if (it != _variables_states.end()) {
+                primitive->set_output_memory(it->second->memory, false);
+            }
+        }
+    }
+}
+
+void network::update_variable_memory(const std::string& variable_id, const cldnn::layout& layout) {
+    auto it = _variables_states.find(variable_id);
+    if (it != _variables_states.end()) {
+        if (it->second->memory->get_layout() != layout) {
+            it->second->set_memory(get_engine().allocate_memory(layout, false));
+            for (auto primitive : _variable_state_primitives) {
+                if (const auto& memory_state_primitive = std::dynamic_pointer_cast<memory_state::variable>(primitive)) {
+                    if (!variable_id.compare(memory_state_primitive->variable_id())) {
+                        primitive->set_output_memory(it->second->memory, false);
```

Comment:
Also, I'd propose to think about changing state implementation a little bit.
As I understand, currently Virables are not a part of graph and stored separately in network. But I think it can be added as a `mutable_data` primitive, so read/assign primitives will be connected with it to guarantee correct processing order.

I assume that with such a change we won't need to iterate over state primitives and update each of them with new mem ptr, you'll just update mutable_data once and then common code will propagate the info.

Reply:
I think empty tensor (or tensor with single element) would be fine.